packages/selected-network-controller/src/SelectedNetworkController.ts

```diff
@@ -167,36 +171,42 @@ export class SelectedNetworkController extends BaseController<
 
     this.update((state) => {
       state.domains[domain] = networkClientId;
-      if (!state.perDomainNetwork) {
-        state.domains[METAMASK_DOMAIN] = networkClientId;
-      }
     });
   }
 
+  #domainHasPermissions(domain: Domain): boolean {
+    return this.messagingSystem.call(
+      'PermissionController:hasPermissions',
+      domain,
+    );
+  }
+
   setNetworkClientIdForDomain(
     domain: Domain,
     networkClientId: NetworkClientId,
   ) {
-    if (!this.state.perDomainNetwork) {
-      Object.entries(this.state.domains).forEach(
-        ([entryDomain, networkClientIdForDomain]) => {
-          if (
-            networkClientIdForDomain !== networkClientId &&
-            entryDomain !== domain
-          ) {
-            this.#setNetworkClientIdForDomain(entryDomain, networkClientId);
-          }
-        },
+    if (domain === METAMASK_DOMAIN) {
+      throw new Error(
+        'NetworkClientId for domain "metamask" cannot be set on the SelectedNetworkController',
       );
     }
+
+    if (!this.#domainHasPermissions(domain)) {
+      throw new Error(
+        'NetworkClientId for domain cannot be called with a domain that has not yet been granted permissions',
+      );
+    }
+
     this.#setNetworkClientIdForDomain(domain, networkClientId);
   }
 
   getNetworkClientIdForDomain(domain: Domain): NetworkClientId {
-    if (this.state.perDomainNetwork) {
-      return this.state.domains[domain] ?? this.state.domains[METAMASK_DOMAIN];
+    const { selectedNetworkClientId: metamaskSelectedNetworkClientId } =
+      this.messagingSystem.call('NetworkController:getState');
+    if (!this.state.perDomainNetwork) {
+      return metamaskSelectedNetworkClientId;
```

Comment:
It seems that when `perDomainNetwork` is false, only this method will work, but the other ones won't. Because we are preventing consumers from using this controller for the globally selected network, it seems that they would have to manage some logic prior to using this controller anyway. For instance, they would have to have some kind of conditional where they use NetworkController for domain "metamask" and this controller otherwise.

Therefore, in order to make this controller more purposeful, should we also throw an error here too? Essentially what we'd be saying is that this controller should _only_ be used for domains that are not "metamask", and we'd enforce it at every step. And then at that point, do we even need `perDomainNetwork`? We could also remove this as well.

Reply:
Yeah I can see that and was thinking the asymmetry here was a bit messy. The main reason why I thought it was ok is that the other methods particularly `getProviderAndBlockTracker` involve returning creating and/or returning proxies for the globally selectedNetworkClientId which would create two entirely separate sets of provider/blockTracker, whereas this method is just forwarding along the state from the `NetworkController`. But yeah happy to make it more uniformly unconcerned with the NetworkController's `selectedNetworkClientId`.

Reply:
Yeah, ultimately it does depend on how consumers end up using this controller. Your changes definitely make this controller safer to use from a state perspective so if that's the main concern here then I guess we can call it good for now and maybe address that in another PR.

Reply:
I don't mind if we leave this here for convenience, particularly in the selected network middleware where I imagine having SelectedNetworkController pass through the NetworkController's selectedNetworkClientId to be helpful. Looks like the MetaMaskController would also benefit from not having to deal with figuring it out if it should be reaching into the SelectedNetworkController or the NetworkController based on the origin value.

Makes sense to leave as is imo